### PR TITLE
weth and eth handling on ethereum

### PR DESCRIPTION
**Pull request by 0xPrimata**
# Summary
- **RFCs**: [315](https://github.com/movementlabsxyz/movement/issues/314)
- **Categories**: `protocol-units`

Adds Enum handling of different assets

# Changelog

- Adds AssetType enum which describes the asset being bridged, future proofing addition of new assets
- Adds methods to handle variants EthAndWeth and Moveth.
- adds eth client and test harness methods to handle WETH
- adds tests to handle both eth and weth

# Testing

Passing tests

# Outstanding issues

Attention to Moveth

### CODE REVIEW

File: protocol-units/bridge/chains/ethereum/src/client.rs
Comment (0xmovses) [inline protocol-units/bridge/chains/ethereum/src/client.rs:255]: let's remove all occurrences of `println!`

File: protocol-units/bridge/chains/ethereum/src/client.rs
Comment (0xmovses) [inline protocol-units/bridge/chains/ethereum/src/client.rs:271]: Remove these `println!`

File: protocol-units/bridge/chains/ethereum/src/client.rs
Comment (0xmovses) [inline protocol-units/bridge/chains/ethereum/src/client.rs:276]: All this logic inside `if amount.weth() > 0 ` I would put inside the integration test. It's not the responsibility of this function to do WETH calls, its just for calling the bridge contracts. 

In real world if a user hasn't called approve on a WETH contract we shouldn't be handling that for them off-chain in a hidden way. 

Put all this logic inside the `eth_movement.rs` integration test, perhaps in a utility function so that you can reuse it. It could even be a method on the `TestHarness`, as its technically a testing utility method.

File: protocol-units/bridge/chains/ethereum/src/client.rs
Comment (0xmovses) [inline protocol-units/bridge/chains/ethereum/src/client.rs:288]: We should remove the underscore from `_initiator_address` as its now being used.

File: protocol-units/bridge/chains/ethereum/src/client.rs
Comment (0xmovses) [inline protocol-units/bridge/chains/ethereum/src/client.rs:368]: Wouldn't the asset always be WETH on `BridgeTransferDetails`? As its what's stored on the contract? No Eth is every stored in the contract.

File: protocol-units/bridge/chains/ethereum/src/client.rs
Comment (0xmovses) [inline protocol-units/bridge/chains/ethereum/src/client.rs:398]: We should not use `unwrap()` in production code. Instead you want to return a `BridgeContractInitiatorResult`

```
U256::try_from(amount.0).map_err(|e| {
        // Handle the error appropriately, possibly converting it to your custom error type
        BridgeContractInitiatorError::ConversionError
    })
```

And you would need to add `ConversionError` variant to `BridgeContractInitiatorError` Because:
`pub type BridgeContractInitiatorResult<T> = Result<T, BridgeContractInitiatorError>;`

https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err

File: protocol-units/bridge/chains/movement/src/lib.rs
Comment (0xmovses) [inline protocol-units/bridge/chains/movement/src/lib.rs:406]: The `Amount` type on Movement shouldn't be `weth()` but `Moveth`, this is why originally I was wanting Amount to have a generic param, so that on Movement side it could just be `u64`, we'll have to solve this just with the enum approach now. 

Probably a method `.moveth()`, even if the inner logic of the method is the same, the semantics are wrong.

File: protocol-units/bridge/integration-tests/tests/eth_movement.rs
Comment (0xmovses) [inline protocol-units/bridge/integration-tests/tests/eth_movement.rs:282]: We're missing the case `test_eth_client_should_successfully_call_initiate_transfer_only_eth`, let's add it.

File: protocol-units/bridge/shared/src/types.rs
Comment (0xmovses) [inline protocol-units/bridge/shared/src/types.rs:142]: Let's add doc comments for AssetType and its fields
```suggestion
/// The type of Asset being used
#[derive(Clone, Debug, PartialEq, Eq, Copy)]
```

File: protocol-units/bridge/shared/src/types.rs
Comment (0xmovses) [inline protocol-units/bridge/shared/src/types.rs:144]:
```suggestion
       /// Where the first tuple value is `Eth` and the second tuple value is `Weth`
	EthAndWeth((u64, u64)),
```

**Review by 0xmovses — CHANGES_REQUESTED** [review]
Thanks for this! Few changes needed.

File: protocol-units/bridge/integration-tests/tests/eth_movement.rs
Comment (0xPrimata) [reply protocol-units/bridge/integration-tests/tests/eth_movement.rs:282]: it's above, it's not named _only_eth

File: protocol-units/bridge/chains/ethereum/src/client.rs
Comment (0xPrimata) [reply protocol-units/bridge/chains/ethereum/src/client.rs:368]: yes, that's why on BridgeTransferDetails we only use the second param for EthAndWeth

File: protocol-units/bridge/chains/movement/src/lib.rs
Comment (0xPrimata) [reply protocol-units/bridge/chains/movement/src/lib.rs:406]: Think that accessing .0 here would be fine, correct?

Comment (0xmovses) [reply protocol-units/bridge/chains/movement/src/lib.rs]: I mean, yes it would work, its just the semantics / method name I'm referring to. It would be confusing for the reader as they would think its actually weth being used here.

**Comment by andygolay** [comment]
Mind filling out the PR description? For anything other than very small PRs we should be using the template provided.

**Comment by 0xmovses** [comment]
Yep lets add some documentation into the PR description, then this is good to merge 🚀

File: protocol-units/bridge/chains/ethereum/src/client.rs
Comment (0xmovses) [inline protocol-units/bridge/chains/ethereum/src/client.rs:406]: If the `expect` is executed, the service will panic causing a crash. https://doc.rust-lang.org/std/result/enum.Result.html#method.expect

Some of these expect calls were left in when initially writing the service, let's remove them now. Instead you could to this, where you have implemented the trait `TryFrom` set`type Error = BridgeContractCounterpartyError`

Because `lock_bridge_transfer_assets` returns `BridgeContractCounterpartyResult<()>` and this is 
`pub type BridgeContractCounterpartyResult<T> = Result<T, BridgeContractCounterpartyError>;`

You should then be able to do `U256::try_from(amount.0)?`

File: protocol-units/bridge/chains/movement/src/lib.rs
Comment (0xmovses) [inline protocol-units/bridge/chains/movement/src/lib.rs:406]: 👍

File: protocol-units/bridge/shared/src/types.rs
Comment (0xmovses) [inline protocol-units/bridge/shared/src/types.rs:182]: Have this return an Result type instead of panicing as this would cause the program to crash. The Result type should be the same of the calling context then when you call `add_assign` you can append it with `?` to bubble up the error.

File: protocol-units/bridge/shared/src/types.rs
Comment (0xmovses) [inline protocol-units/bridge/shared/src/types.rs:146]: I know we've gone over this a few times, but I still don't understand why we can't have:
```rust
pub enum AssetType {
   Eth(u64)
   Weth(u64)
   EthAndWeth((u64, u64))
   Moveth(u64)
}
```

Otherwise we have code like 
`Amount(AssetType::EthAndWeth((1000,0))),` 

This reads weird because its not Eth and Weth, it's just Eth. Do you see what I mean?

**Review by 0xmovses — CHANGES_REQUESTED** [review]
Looking good! Some room for a bit more improvement.

File: protocol-units/bridge/shared/src/types.rs
Comment (0xPrimata) [reply protocol-units/bridge/shared/src/types.rs:146]: Yep, but then if we are working with the bridge service we have to handle unexpected types. Notice that we cannot use more than one type. Also, this streamlines eth and weth handling for any call to the eth side. What if we expect eth only but get weth too? That weth is lost in value. I just feel like and eth and weth variants is over engineering. We can call ethandweth something else though

Comment (0xmovses) [reply protocol-units/bridge/shared/src/types.rs:146]: In rust any unexpected type at runtime would have to be `dyn` which its not here. These are all concrete types known at compile time. 

Any new `AssetType` could just be added as an enum variant, for eg `Btc`, `Sol`, `Usdc` etc. We can add as many as we want. So it becomes code as documentation. This would setup a nice pattern for configuration and extension. 

In your example, if we expect only eth but get weth too that would match on the `EthAndWeth` arm, and we handle that case.  I don't see the problem and see it only as a worthwhile improvement that improves readability, extensibility (more asset types) and semantics.   

Anything coming "out" from ETH initiator is always WETH.

Comment (0xPrimata) [reply protocol-units/bridge/shared/src/types.rs:146]: But then if we are handling BTC, SOL, USDC, etc, they would be new enum types, for sure, but they are not being handled as BTC and WBTC, they'll be just a single type. If we are handling BTC and WBTC from a chain that supports native and wrapped btc (Rootstock for example), We add a variant BtcAndWbtc((u64,u64)). This way it's easy to be explicit that this implementation, on this side of the bridge, handled two assets for the sake of compatibility with the chain.

File: protocol-units/bridge/chains/movement/src/lib.rs
Comment (0xmovses) [inline protocol-units/bridge/chains/movement/src/lib.rs:462]: I think you resolved the conflict incorrectly, use the new utils:: methods created that bubble up the error with ? Don't unwrap. 

These were added by @andygolay in https://github.com/movementlabsxyz/movement/pull/215/commits/3cde18ff89de6330d620b3a0de7676632ea1caa4

**Review by 0xmovses — CHANGES_REQUESTED** [review]
Wrong change accepted on the merge conflict. To avoid this all three of us @0xPrimata @andygolay and me should all be reviewing each others PR. 

Unless the PR is time-sensitive let's create a soft-rule that we want to approvals form other bridge Team members before approval. This will help create more awareness about eachothers changes and why they were made etc.

File: protocol-units/bridge/chains/movement/src/lib.rs
Comment (0xPrimata) [reply protocol-units/bridge/chains/movement/src/lib.rs:462]: conflict was around line 406, not this one, let me take a look

**Review by 0xmovses — APPROVED** [review]
:shipit:

**Pull request merged**
